include/kframework/lemmas/bytes-simplification.k

```diff
@@ -7,88 +7,240 @@ module BYTES-SIMPLIFICATION [symbolic]
+
+    rule [range-writeRange-before]:
+      #range(#writeRange(B1:Bytes, S1:Int, B2:Bytes), S2:Int, W2:Int) =>
+        #let W = minInt(W2, S1 -Int S2) #in
+          #range(B1, S2, W) +Bytes #range(#writeRange(B1, S1, B2), S1, W2 -Int W)
+      requires S2 <Int S1
+      [simplification]
+
+    rule [range-writeRange-in-between]:
+      #range(#writeRange(B1:Bytes, S1:Int, B2:Bytes), S2:Int, W2:Int) =>
+        #let S = S2 -Int S1 #in
+        #let W = minInt(W2, lengthBytes(B2) -Int S) #in
+          #range(B2, S, W) +Bytes #range(B1, S1 +Int lengthBytes(B2), W2 -Int W)
+      requires S1 <=Int S2 andBool S2 <=Int S1 +Int lengthBytes(B2)
+      [simplification]
+
+    rule [range-writeRange-outside]:
+      #range(#writeRange(B1:Bytes, S1:Int, _:Bytes ), S2:Int, W2:Int) => #range(B1, S2, W2)
+      requires 0 <=Int S1 andBool 0 <=Int W2 andBool S1 +Int lengthBytes(B1) <=Int S2 [simplification]
+
+    // #writeRange
+
+    rule [writeRange-is-empty]:
+      #writeRange(_:Bytes, S:Int, _:Bytes) => .Bytes
+      requires S <Int 0
+      [simplification]
+
+    rule [writeRange-as-concat-inside]:
+      #writeRange(B1, S, B2:Bytes) =>
+        #range(B1, 0, S) +Bytes B2 +Bytes #range (B1 , S +Int lengthBytes(B2), lengthBytes(B1) -Int ( S +Int lengthBytes(B2)))
+      requires 0 <=Int S andBool S +Int lengthBytes(B2) <Int lengthBytes(B1)
+      [simplification(60)]
+
+    rule [writeRange-as-concat-outside]:
+      #writeRange(B1, S, B2:Bytes) => #range(B1, 0, S) +Bytes B2
+      requires 0 <=Int S andBool lengthBytes(B1) <=Int S +Int lengthBytes(B2)
+      [simplification(60)]
+
+    rule [writeRange-reorder]:
+      #writeRange(#writeRange(B:Bytes, S1:Int, B1:Bytes), S2:Int, B2:Bytes) =>
+        #writeRange(#writeRange(B, S2, B2), S1, B1)
+      requires S2 <Int S1 andBool S2 +Int lengthBytes(B2) <=Int S1
+      [simplification]
+
+    rule [writeRange-subsume]:
+      #writeRange(#writeRange(B:Bytes, S1:Int, B1:Bytes), S2:Int, B2:Bytes) => #writeRange(B, S2, B2)
+      requires S2 <=Int S1 andBool S1 +Int lengthBytes(B1) <=Int S2 +Int lengthBytes(B2)
+      [simplification]
 
-    rule #range(#writeRange(M, N, BUF:Bytes ), L, K) => #range(M, L, K)
-      requires 0 <Int K andBool N +Int lengthBytes(BUF) <=Int L [simplification]
+    // lengthBytes
 
-    rule #range(#writeRange(#writeRange(M, N1, BUF1:Bytes), N2, BUF2:Bytes), L, K ) => #range(#writeRange(M, N2, BUF2), L, K)
-        requires 0 <Int K andBool N1 +Int lengthBytes(BUF1) <=Int L [simplification]
+    rule [lengthBytes-geq-zero]    : 0 <=Int lengthBytes(_:Bytes)                       => true                                           [simplification, smt-lemma]
+    rule [lengthBytes-concat]      : lengthBytes(B1:Bytes +Bytes B2:Bytes)              => lengthBytes(B1) +Int lengthBytes(B2)           [simplification]
+    rule [lengthBytes-buf]         : lengthBytes(#buf(W:Int, _:Int))                    => W requires 0 <=Int W                           [simplification]
```

Comment:
This could write a potentially undefined term to a defined one. We should have the extra condition `#rangeUInt(8 *Int W, I)`

Reply:
Hmm, but `#buf` is total?

Reply:
Actually, `#buf(W, D)` is defined for any `D`, it calculates a remainder with respect to `2 ^ (8 *Int W)`, so it should be fine.